osu.Game/Online/LocalUserStatisticsProvider.cs

```diff
@@ -0,0 +1,94 @@
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections.Generic;
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Game.Extensions;
+using osu.Game.Online.API;
+using osu.Game.Online.API.Requests;
+using osu.Game.Rulesets;
+using osu.Game.Users;
+
+namespace osu.Game.Online
+{
+    /// <summary>
+    /// A component that is responsible for providing the latest statistics of the logged-in user for the game-wide selected ruleset.
+    /// </summary>
+    public partial class LocalUserStatisticsProvider : Component
+    {
+        /// <summary>
+        /// The statistics of the logged-in user for the game-wide selected ruleset.
+        /// </summary>
+        public IBindable<UserStatistics?> Statistics => statistics;
+
+        private readonly Bindable<UserStatistics?> statistics = new Bindable<UserStatistics?>();
+
+        [Resolved]
+        private IBindable<RulesetInfo> ruleset { get; set; } = null!;
+
+        [Resolved]
+        private IAPIProvider api { get; set; } = null!;
+
+        private readonly Dictionary<string, UserStatistics> allStatistics = new Dictionary<string, UserStatistics>();
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            statistics.BindValueChanged(v =>
+            {
+                if (api.LocalUser.Value != null && v.NewValue != null)
+                    api.LocalUser.Value.Statistics = v.NewValue;
+            });
+
+            ruleset.BindValueChanged(_ => updateStatisticsBindable());
+
+            api.LocalUser.BindValueChanged(_ =>
+            {
+                allStatistics.Clear();
+                updateStatisticsBindable();
+            }, true);
+        }
+
+        private GetUserRequest? currentRequest;
+
+        private void updateStatisticsBindable() => Schedule(() =>
+        {
+            statistics.Value = null;
+
+            if (api.LocalUser.Value == null || api.LocalUser.Value.OnlineID <= 1 || !ruleset.Value.IsLegacyRuleset())
+            {
+                statistics.Value = new UserStatistics();
+                return;
+            }
+
+            if (currentRequest?.CompletionState == APIRequestCompletionState.Waiting)
+            {
+                currentRequest.Cancel();
+                currentRequest = null;
+            }
+
+            if (allStatistics.TryGetValue(ruleset.Value.ShortName, out var existing))
+                statistics.Value = existing;
+            else
+                requestStatistics(ruleset.Value);
+        });
+
+        private void requestStatistics(RulesetInfo ruleset)
+        {
+            currentRequest = new GetUserRequest(api.LocalUser.Value.OnlineID, ruleset);
+            currentRequest.Success += u => statistics.Value = allStatistics[ruleset.ShortName] = u.Statistics;
+            api.Queue(currentRequest);
+        }
```

Comment:
I did not emphasise too much about this, but I have mentioned in OP the following:
> This could be improved by fetching statistics of all rulesets once the user is logged in, but it would require performing four consecutive GetUserRequests for each ruleset to provide the statistics in full detail (i.e. when performing a GetUsersRequest, the RulesetsStatistics dictionary does not provide information about the country ranking in each ruleset).

`SoloStatisticsWatcher` is initialising user statistics for all rulesets using a `GetUsersRequest`, the response of that request does not include country ranking and potentially other attributes.

If anything, I would foresee this new component exposing statistics for all rulesets by a method or a dictionary, and `DifficultyRecommender` turning into an extension method, and `SoloStatisticsWatcher` becoming completely dependent on `LocalUserStatisticsProvider` and do away with the API request logic in the watcher. 

However, right now it's sort-of not pretty to achieve that, as I have to make the component perform one-time four consecutive API requests to receive full user statistics on each ruleset, and in complete honestly, I have no idea whether this is something that you would be fine with or not.

Reply:
at this point i'd probably ask if it is possible for us to receive full statistics in a single request because this feels wasteful.

@ppy/team-web do you think we could change some user-fetching endpoint to return complete rank data, including country ranking?

Reply:
country ranking is rather expensive

Reply:
> both `SoloStatisticsWatcher` and `DifficultyRecommender` do initial population of statistics for their own uses for _all rulesets_. so why is this doing the same in a _third_ place?
> 
> this is why i was floating potentially exposing the internal dictionary from `SoloStatisticsWatcher`.

Looking back at this, since the purpose of `LocalUserStatisticsProvider` is to track latest statistics and provide the latest one, I've made `UserStatisticsWatcher` fully rely on it and remove its local part of fetching initial statistics and tracking statistics in general.